### DISCUSSION TITLE
Trainer not optimizing weights

### DISCUSSION THREAD

**Discussion by cwoolfo1** [Lightning-AI/pytorch-lightning#10684] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
Hello, I am running code to train a stacked GRU. I have defined and created train and validation data loaders. I initialized the trainer but when I run the code the accuracies for the validation loop and train loop don't change, they stay the same. I have compared my code to the documentation and I cannot seem to find the issue.

```
class DataModule(p1.LightningDataModule):
    def __init__(self):
        super().__init__()
        self.input = np.load('4thin.npy')
        self.output = np.load('4thout.npy')
        # Loads in npy files
        self.input = self.input[:,:,:10]
        self.output = self.output[:,:,:8]
        self.output = self.output.squeeze()
        self.batch_size = 1
        self.num_workers = 1
        # Set number of samples per batch
        
    def setup(self, stage):
        if stage == 'fit':
            split = len(self.input[:,1,1]) / 2
            split = int(split)
            self.trainX, self.trainy = self.input[:split, :, :], self.output[:split, :]
            self.valX, self.valy = self.input[split:, :, :], self.output[split:, :]
            # Splits the dataset into two smaller datasets for training.
            
        if stage == 'test' or stage is None:
            self.testX, self.testy = self.input, self.output
            # Returns the full dataset for testing.
            
    def train_dataloader(self):
        data = self.trainX, self.trainy
        return DataLoader(CustomDataset(data), batch_size = self.batch_size, num_workers=self.num_workers)
    # Loads data for training.
    
    def val_dataloader(self):
        data = self.valX, self.valy
        return DataLoader(CustomDataset(data), batch_size = self.batch_size, num_workers=self.num_workers)
    # Loads data for validation.
    
    def test_dataloader(self):
        data = self.testX, self.testy
        return DataLoader(CustomDataset(data), batch_size = self.batch_size, num_workers=self.num_workers)
    # Loads data for testing.        

parser = ArgumentParser()
parser.add_argument("--gpus", default=0)
parser.add_argument("--num_nodes", default=1)
args = parser.parse_args()
dm = DataModule()
dm.setup('fit')
train_dataloader = dm.train_dataloader()
val_dataloader = dm.val_dataloader()
model = NeuralNetwork(10, 10, 8, 2,)
# Creates a Stacked GRU model of the form (feature size, hidden size, output size, # of GRU layers)
tb_logger = p1_loggers.TensorBoardLogger("logs/")
# Defines the logger as a tensor board logger for training.
n = 1
trainer = Trainer(num_nodes = 1, gpus = args.gpus, logger=tb_logger, accumulate_grad_batches=1, reload_dataloaders_every_n_epochs=n) #, strategy="ddp")
# Uses tensorboard to log metrics and gradient accumulation to speed up processes.
trainer.fit(model, train_dataloader, val_dataloader)
```

Attached is the code I used to define my data module class and initialize and train my model. Is there any advice y'all can offer me for fixing this issue?

**Comment by rohitgr7** [comment]
can you share the code for `NeuralNetwork`?

**Reply by cwoolfo1** [reply]
```
class NeuralNetwork(LightningModule):
    def __init__(self, input_size, hidden_size, output_size, num_layers):
        super(NeuralNetwork, self).__init__()
        self.automatic_optimization = True
        self.input_size = input_size
        self.hidden_size = hidden_size
        self.output_size = output_size
        self.num_layers = num_layers
        # Above initializes all the variables.
        self.hiddenlayers = nn.GRU(input_size, hidden_size, num_layers=num_layers, batch_first=True)
        self.outputlayer = nn.Linear(input_size, output_size)
        # This code will create a stacked GRU with n number of layers.
    def forward(self, x):
        x = x.float()
        output, hidden = self.hiddenlayers(x)
        x = F.relu(hidden[-1])
        # Pushes input through the stacked GRU.
        # Returns the last GRU layers last hidden state.
        x = F.softmax(self.outputlayer(x), dim=1)
        # Pushes that hidden state through a linear softmax layer.
        return x
    def training_step(self, batch, batch_idx):
        x, y = batch
        y = y.float()
        # Typecast to float tensor
        logits = self(x)
        # Run through neural network
        loss = torch.nn.MSELoss()
        # Mean squared error loss
        loss = loss(logits, y)
        self.log("train_loss", loss, prog_bar=True, logger=True)
        # Log the loss for each epoch.
        return loss
    def validation_step(self, batch, batch_idx):
        x, y = batch
        y = y.float()
        # Typecast to float tensor
        logits = self(x)
        # Run through neural network
        loss = torch.nn.MSELoss()
        loss = loss(logits, y)
        # Change to cross entropy loss
        preds = torch.argmax(logits, dim=0)
        acc = accuracy(preds, y.int())
        # Typecast to int because accuracy needs int target
        # Calling self.log will surface up scalars for you in TensorBoard
        self.log("val_loss", loss, prog_bar=True, logger=True)
        self.log("val_acc", acc, prog_bar=True, logger=True)
        return loss

    def test_step(self, batch, batch_idx):
        # Here we just reuse the validation_step for testing
        return self.validation_step(batch, batch_idx)

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=.0003)
        return optimizer
```

**Reply by rohitgr7** [reply]
you are doing a softmax here:
```
x = F.softmax(self.outputlayer(x), dim=1)
```
and passing the probabilities to mseloss. is this correct?

**Reply by cwoolfo1** [reply]
Yes

**Reply by rohitgr7** [reply]
your code looks okay to me.. not sure what's causing the issue. is training and validation loss reducing with time?

**Reply by cwoolfo1** [reply]
No it is not. It goes from epoch to epoch but the training and validation loss do not change.

**Reply by rohitgr7** [reply]
oh! can you compile a small reproducible example in a colab notebook? I'll check why is it not training :)
